package.json

```diff
@@ -3,14 +3,20 @@
   "version": "1.0.0",
   "description": "A bridge between Matrix and Discord",
   "main": "discordas.js",
+  "engines": {
+    "npm": "please-use-yarn",
+    "yarn": "^1.22.0",
+    "node": "12.x - 16.x"
+  },
   "scripts": {
+    "getyarn": "yarn set version 1.22",
```

Comment:
I'm ever so worried with locking yarn versions, because we will only forget to deal with this later. What's the rationale for minor version locking?

Reply:
Rationale is to prevent breakage by a later version of Yarn 1, but given that v1 is "stable" now, such breakage shouldn't be a threat.  Locking on any 1.x should be fine; what really matters is to prevent using Yarn >=2.

Reply:
Never mind: while Yarn 1 runs preinstall scripts before resolving packages, Yarn 2+ runs them _after_ it, which I've verified experimentally & [is documented](https://yarnpkg.com/advanced/lifecycle-scripts):

> **postinstall** is called after a package's dependency tree changes are written to the disk...**preinstall** and **install** are called as part of **postinstall**.

See also:
- https://github.com/yarnpkg/berry/issues/2689
- https://github.com/yarnpkg/berry/issues/2701

As such, there doesn't seem to be a way to make `yarn` run a version-check script (or any script) before/at the start of the install step in Yarn 2+.

This means my running my `getyarn` script via `yarn getyarn` won't help either, because Yarn 2+ can't run it until after a successful `yarn`/`yarn install`--which fails on this project, thus defeating the purpose of the script.

Yarn does support enforcing a version via the `"packageManager"` key in `package.json`...but it's supported only Yarn 2+.

So [checking in the Yarn version to use](https://github.com/yarnpkg/berry/discussions/3064) seems to be the only reliable way to enforce Yarn 1 after all (and [`yarn policies set-version`](https://classic.yarnpkg.com/lang/en/docs/cli/policies/) exists to help set that up).  With that said, checking in a version is even more strict than just locking which version to download...

At this point, I'm inclined to give up programmatically enforcing Yarn 1 & just recommend using `yarn set version classic` as a manual setup step.  If/when we decide to move to Yarn 2+, these problems will be a thing of the past.

Reply:
Yeah, I think there is significant value in just mentioning this in the README, which people can be referred to if they get stuck, imo.